### PR TITLE
Overview diff page

### LINKED ISSUE #42: Overview Diff Page
will show the diff of two overview scans

Screen Shot 2020-02-12 at 16 23 56

### PR DESCRIPTION
**Pull request by detaybey**
**What this PR does / why we need it**:

- The _InfrastructureComponentSummary_ now has a component property.
- New route (overviewdiff) added on API, which takes two dates and serves an _InfrastructureOverviewDiff_
- The mock data structure was changed, static functions _GetCounters()_,  _GetScoreHistory()_, _GetComponentSummary()_ and _GetComponentSummaries()_ were added to return nested date/component/counter data (to reduce the number of repetitive copy-paste dummy data).

**Which issue(s) this PR fixes**:
Fixes #42

### CODE REVIEW

File: src/backend/joseki.be/webapp/Models/InfrastructureOverviewDiff.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{
+    /// <summary>
+    /// Returns infrastructure summary: scores by category, counters, etc.
+    /// </summary>
+    public class InfrastructureOverviewDiff
+    {
+        /// <summary>
+        /// First overall infrastructure summary.
+        /// </summary>
+        public InfrastructureComponentSummary Overall1 { get; set; }
```
Comment (v1r7u) [inline src/backend/joseki.be/webapp/Models/InfrastructureOverviewDiff.cs]: I guess, for `diff` view, you do not need `ScoreHistory` and `ScoreTrend`, that are part of `InfrastructureComponentSummary`. Maybe it makes sense to add another data-type for this?

File: src/backend/joseki.be/webapp/Controllers/AuditsController.cs [hunk 2]
```diff
@@ -15,32 +16,54 @@ namespace webapp.Controllers
...
+
+            return Task.FromResult(this.StatusCode(200, summary));
+        }
+
+        /// <summary>
+        /// Returns the overall infrastructure overview diff.
+        /// </summary>
+        /// <returns>The overall infrastructure overview diff.</returns>
+        [HttpGet]
+        [Route("overviewdiff", Name = "get-overview-diff")]
```
Comment (v1r7u) [inline src/backend/joseki.be/webapp/Controllers/AuditsController.cs]: What about a route `/api/overview/diff?date1=20200101&date2=20200102`, where `date1` and `date2` are query parameters? I think, it's more REST-friendly

File: src/backend/joseki.be/webapp/Controllers/AuditsController.cs [hunk 3]
```diff
@@ -15,32 +16,54 @@ namespace webapp.Controllers
...
+        [ProducesResponseType(200, Type = typeof(InfrastructureOverviewDiff))]
+        public Task<ObjectResult> GetOverviewDiff(DateTime? date1 = null, DateTime? date2 = null)
```
Comment (v1r7u) [inline src/backend/joseki.be/webapp/Controllers/AuditsController.cs]: Likely, we can make dates non-nullable

File: src/backend/joseki.be/webapp/Models/InfrastructureComponentSummary.cs [hunk 4]
```diff
@@ -8,19 +8,19 @@ namespace webapp.Models
     public class InfrastructureComponentSummary
     {
         /// <summary>
-        /// The unique Id of a component.
+        /// The unique Id of a component summary.
         /// </summary>
         public string Id { get; }
```
Comment (v1r7u) [inline src/backend/joseki.be/webapp/Models/InfrastructureComponentSummary.cs]: Probably, we do not need an Id here anymore

File: src/backend/joseki.be/webapp/Controllers/AuditsController.cs [hunk 5]
```diff
@@ -15,32 +16,54 @@ namespace webapp.Controllers
     [Route("api/audits")]
     public class AuditsController : Controller
```
Comment (v1r7u) [inline src/backend/joseki.be/webapp/Controllers/AuditsController.cs]: A general proposal.
Dates are often represented in different ways on different platforms (misc to-string serialization rules with timezone and without, unix-epoch, etc). 
What about exposing _dates_ from REST api as strings with a particular format (for example, `YYYYMMDD`) and then parsing it to `DateTime` object inside this controller methods?

File: src/backend/joseki.be/webapp/Models/InfrastructureOverviewDiff.cs (hunk 1 above)
Comment (detaybey) [reply src/backend/joseki.be/webapp/Models/InfrastructureOverviewDiff.cs]: now we have `InfrastructureComponentSummary` and `InfrastructureComponentSummaryWithHistory` for this purpose.

File: src/backend/joseki.be/webapp/Controllers/AuditsController.cs (hunk 2 above)
Comment (detaybey) [reply src/backend/joseki.be/webapp/Controllers/AuditsController.cs]: makes sense! route changed to `/api/overview/diff`

File: src/backend/joseki.be/webapp/Controllers/AuditsController.cs (hunk 3 above)
Comment (detaybey) [reply src/backend/joseki.be/webapp/Controllers/AuditsController.cs]: done!

File: src/backend/joseki.be/webapp/Models/InfrastructureComponentSummary.cs (hunk 4 above)
Comment (detaybey) [reply src/backend/joseki.be/webapp/Models/InfrastructureComponentSummary.cs]: removed the `Id` for `InfrastructureComponentSummary`

File: src/backend/joseki.be/webapp/Controllers/AuditsController.cs (hunk 5 above)
Comment (detaybey) [reply src/backend/joseki.be/webapp/Controllers/AuditsController.cs]: definitely, we should decide on a format and pass around strings. but question; what would happen if two scans run on same day? i remember we discussed this but don't recall we had a resolution.  I still believe we should keep the time portion.

Comment (v1r7u) [reply src/backend/joseki.be/webapp/Controllers/AuditsController.cs]: We agreed to persist the latest scan per day only

**Pull request merged**